---
title: "Exercise Five"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(DESeq2)
library(tidyverse)
#remotes::install_github("compgenomr/compGenomRData")
# uncomment the line above if you DON'T already have this package installed
```

For this exercise, we will be using a subset of count data from a colorectal cancer study, as described here - 
http://compgenomr.github.io/book/gene-expression-analysis-using-high-throughput-sequencing-technologies.html#computing-different-normalization-schemes-in-r

First run all the code in the chunk below. This code prepares the `countData` and `colData` objects we will need to make a DESeqDataSet object  

Run *all* of the code in the chunk below before attempting the subsequent questions

```{r}
counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv",
                           package = "compGenomRData")
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv",
                            package = "compGenomRData")

counts <- as.matrix(read.table(counts_file, header = T, sep = '\t'))

#remove the 'width' column
countData <- as.matrix(subset(counts, select = c(-width)))
#define the experimental setup 
colData <- read.table(coldata_file, header = T, sep = '\t', 
                      stringsAsFactors = TRUE)

# run DESeq
dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = ~ group) 
dds <- DESeq(dds)
res <- results(dds,
           contrast = c("group", "CASE", "CTRL"))
```

Q1. Make an MA-plot using the `res` object. What does this plot tell you? What do the blue dots indicate
A1. 
```{r}

```

Q2. Convert `res` to a dataframe called `res_df`. Retrieve the Entrez IDs for the gene symbols in `res_df``
A2.
```{r}
library("biomaRt")


```

Q3. How many `NAs` are there in the `entrez_id` column of `res_df`? Remove these `NAs` from `res_df``
A3.
```{r}

```

Q4. Are all the Entrez IDs in `res_df` unique? Drop any repeats
A4.
```{r}

```

Q5. Perform over-representation analysis using only genes with an adjusted p-value < 0.05, and the "CC" ontology. What is the most significant term?
A5.
```{r}
library(clusterProfiler)
library(org.Hs.eg.db)


```







