---
title: "Class Five (Answers): Computational Biology Reading Group"
author: "Your name"
date: "Nov 17, 2021"
output: html_document
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Install CRAN packages
installifnot <- function(pckgName){
  if (!(require(pckgName, character.only = TRUE))) {
    install.packages(pckgName, dep = TRUE)
    require(pckgName, character.only = TRUE)
  }
}

cran_pkgs <- c("BiocManager", "remotes", "tidyverse")

for (i in 1:length(cran_pkgs)){
  installifnot(cran_pkgs[i])
}

# Install Bioconductor packages
installBiocifnot <- function(pckgName){
  if (!(require(pckgName, character.only = TRUE))) {
    BiocManager::install(pckgName)
    require(pckgName, character.only = TRUE)
  }
}

bioc_pkgs <- c("DESeq2", "UCLouvain-CBIO/rWSBIM2122", "clusterProfiler", "biomaRt", "org.Hs.eg.db", "enrichplot", "fgsea")

for (i in 1:length(bioc_pkgs)){
  installBiocifnot(bioc_pkgs[i])
}

# Load packages
library(rWSBIM2122)
library(DESeq2)
library(tidyverse)
library(clusterProfiler)
library(biomaRt)
library(org.Hs.eg.db)
library(enrichplot)
library(fgsea)
```

# Performing RNA-seq analysis using the DESeq2 package - Enrichment Analaysis

For this walk-through we will be using the same example (and much of the same code!) as in this chapter - 
https://uclouvain-cbio.github.io/WSBIM2122/sec-gsea.html

Refer to these guides for more information on how to use -
clusterProfiler - https://yulab-smu.top/biomedical-knowledge-mining-book/enrichment-overview.html
fgsea - http://bioconductor.org/packages/release/bioc/vignettes/fgsea/inst/doc/fgsea-tutorial.html
msigdb gene sets - https://data.broadinstitute.org/gsea-msigdb/msigdb/release/

Wonderful blogpost on using the fgsea package - https://stephenturner.github.io/deseq-to-fgsea/


# Load data and perform differential expression analysis

```{r load-and-run}
# system.file() finds the location of the specified file in the specified package
# load() then opens the specified file (`.rda` objects in this case) so that they appear in our global environment 
load(system.file("extdata/deseq2/counts.rda",
                  package = "rWSBIM2122"))
load(system.file("extdata/deseq2/coldata.rda",
                  package = "rWSBIM2122"))
# you should now see two new objects in your Environment pane (top right-hand corner)

# Construct a DESeqDataSet object
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = coldata,
                              design = ~ Condition) 

# Run DESeq2
dds <- DESeq(dds)

res <- results(dds,
               contrast = c("Condition", "KD", "mock")) # we do this so that `mock` becomes the group we're comparing to
```

# Shrinking log fold-changes for low-count genes
```{r}
plotMA(res)

resultsNames(dds)
# Output - 
# "Intercept"            "Condition_mock_vs_KD"

# How do we switch the order of the second coefficient so that this reads "Condition_KD_vs_mock"
coldata$Condition <- factor(coldata$Condition, levels = c("mock", "KD"))

# Repeat DESeq steps
# Construct a DESeqDataSet object
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = coldata,
                              design = ~ Condition) 
# Run DESeq2
dds <- DESeq(dds)

resultsNames(dds)
# Output - 
# "Intercept"            "Condition_KD_vs_mock"

res_shrunk <- lfcShrink(dds, 
                        coef = "Condition_KD_vs_mock",
                        type = "apeglm") 
# Here's a nice article on empirical Bayes estimation - http://varianceexplained.org/r/empirical_bayes_baseball/

res_shrunk 
# Compare the output of - 
plotMA(res)
# and
plotMA(res_shrunk)

# Convert `res_shrunk` to a dataframe called `res_df`. We will use this for enrichment analyses
res_df <- as_tibble(res_shrunk, rownames = "ENSEMBL")
```

# Over-representation Analysis (ORA)
```{r}
# Get ENTREZ IDs
# We'll need this for using the "org.Hs.eg.db" package
# Step One: Connect to the selected BioMart database and dataset hosted by Ensembl
ensembl <- useEnsembl(biomart = "genes", 
                   dataset = "hsapiens_gene_ensembl")

# Step Two: Retrieve gene names
# build a biomaRt query
# The getBM() function is the main query function in biomaRt
ensembl_to_entrez <- getBM(attributes = c("ensembl_gene_id", "external_gene_name",
                                            "entrezgene_id"),
                             values = res_df$ENSEMBL,
                             mart = ensembl)

# Plan B in case there's a connection problem -
# ensembl_to_entrez <- read_csv("https://raw.githubusercontent.com/dukecct/CBRG/main/inst/data/ensembl_to_entrez.csv")

# add this new info to res_df
res_df <- res_df %>% 
            left_join(ensembl_to_entrez, by = c("ENSEMBL" = "ensembl_gene_id"))

# remove rows with NAs in the columns in `entrezgene_id` and `padj`
res_df <- res_df %>% 
            drop_na(entrezgene_id, padj)

# are all ENTREZ IDs unique?
length(unique(res_df$entrezgene_id)) # 13711
nrow(res_df) # 13730

# drop duplicates
res_df <- res_df %>% 
            arrange(padj) %>% 
            distinct(entrezgene_id, .keep_all = TRUE)

# are all ENTREZ IDs unique?
length(unique(res_df$entrezgene_id)) == nrow(res_df) # TRUE

# perform ORA
# we need a vector of ENTREZ IDs for genes with padj values < 0.05
sig_genes <- res_df %>% 
              filter(padj < 0.05) %>% 
              pull(entrezgene_id)
sig_genes

go_ora <- enrichGO(gene = sig_genes,
                   OrgDb = org.Hs.eg.db,
                   #universe = as.character(res_df$entrezgene_id),
                   ont = "MF",
                   readable = TRUE) # maps gene IDs to gene names
head(go_ora)

# Visualization
# dot plot
go_ora %>% 
  dotplot(showCategory = 30) + 
  ggtitle("dotplot for ORA")

# heatplot
heatplot(go_ora, showCategory = 5)

# upsetplot
upsetplot(go_ora)
```

# Gene-set enrichment analysis (GSEA)
```{r}
for_gsea <- res_df %>% 
              arrange(desc(log2FoldChange)) %>% 
              dplyr::select(entrezgene_id, log2FoldChange)

ranks <- deframe(for_gsea)
head(ranks)

# Load the pathways into a named list
pathways.hallmark <- gmtPathways("data/h.all.v7.4.entrez.gmt.txt") # you can download this file from https://data.broadinstitute.org/gsea-msigdb/msigdb/release/7.4/

pathways.hallmark

# run fgsea with 1000 permutations
fgseaRes <- fgsea(pathways = pathways.hallmark, 
                  stats = ranks, 
                  nperm=1000)

# tidy results
fgseaResTidy <- fgseaRes %>%
                    as_tibble() %>%
                    arrange(desc(NES))

# plot normalized enrichment scores
ggplot(fgseaResTidy, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill = padj<0.05)) + # "implicit" if-else statement
  coord_flip() +
  labs(x = "Pathway", y = "Normalized Enrichment Score",
       title = "Hallmark pathways NES from GSEA") + 
  theme_minimal()

# enrichment plots for specific pathways
# a highly upregulated pathway
plotEnrichment(pathways.hallmark$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION, 
               ranks)

# an uninteresting pathway
plotEnrichment(pathways.hallmark$HALLMARK_PI3K_AKT_MTOR_SIGNALING, 
               ranks)

# a highly downregulated pathway
plotEnrichment(pathways.hallmark$HALLMARK_MITOTIC_SPINDLE, 
               ranks)

# plot GSEA table
pathways <- fgseaResTidy %>% 
                  dplyr::filter(padj < 0.05 & NES > 0) %>% 
                  pull(pathway)

plotGseaTable(pathways.hallmark[pathways], 
              ranks,
              fgseaRes)
```

