---
title: "Handling Strings in R"
author: "Your name"
date: "Apr 06, 2022"

output:
  html_document: 
    
    toc: true 
    toc_float: 
      collapsed: false 
      smooth_scroll: false 
      
    number_sections: true 
    
    theme: paper 
    
    output: 
      fig_width: 10 
      fig_height: 8 
      fig_caption: true 
      dev: svg 
    
    #Dataframe printing
    df_print: paged 
    
    code_folding: show 
  
  pdf_document: default
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)

cran_pkgs <- c(
  "remotes",
  "janitor",
  "tidyverse"
  )

# Install and load CRAN packages
installifnot <- function(pckgName){
    if (!(require(pckgName, character.only = TRUE))) {
        install.packages(pckgName, dep = TRUE)
        library(pckgName, character.only = TRUE)
    }
}

lapply(cran_pkgs, installifnot)

remotes::install_github("matthewhirschey/tidybiology", force = TRUE)
library(tidybiology)
data(proteins)
```

# Definition and useful references

A "string" in R is anything within quotation marks e.g. "I am a string"

Much of this walkthrough is modeled on the "Strings" of Hadley Wickham's R4DS book - 
https://r4ds.had.co.nz/strings.html

We will mostly be making use of the {stringr} package for this walkthrough. Useful documentation and examples can be found here - https://stringr.tidyverse.org

And here's a handy cheatsheet that describes how the functions in this package work - 
https://evoldyn.gitlab.io/evomics-2018/ref-sheets/R_strings.pdf

# Simple string handling

```{r}
string <- "MSMLVVFLLLWGVTWGPVTEAAIFYETQPSL"

# How many amino acids are in this sequence?
length(string)
str_length(string)
# Why do these functions give different answers?
?str_length

# How many lysines does this sequence contain?

str_view_all(string, "L")

# How many times do two consecutive Alanines occur in this sequence?

str_view_all(string, "AA")

# Does this sequence contain a Proline followed by a Valine followed by a Threonine?

str_view_all(string, "PVT")

# Does this sequence contain a Tryptophan, followed by ANY amino acid, and then a Valine?


```

# Handling strings in Data frames

```{r}
# Take a look at the `proteins` df


# How would you create a new dataframe that only contains proteins that contain the "RCP.L" motif?


# Identify proteins that contain the E-box motif - "CANNTG". Create a new column called `e_box` that contains a `TRUE` if a sequence has this motif and a `FALSE` otherwise


# What are the mean protein lengths for proteins that contain an E-box versus proteins that do not?

```

# Regular expressions

What is a regular expression?

From Wikipedia - A regular expression (shortened as regex or regexp) is a sequence of characters that specifies a search pattern in text

Here's a wonderful (and short!) introduction to Regular Expressions by Roger Peng - https://www.youtube.com/watch?v=NvHjYOilOf8&t=892s

```{r}
# Build two small data frames for these examples
protein_small <- tibble(
protein_names = c("A1BG_HUMAN", "A1CF_HUMAN", "B3GAT1_HUMAN"),
uniprot_id = c("P04217", "Q9NQ94", "Q9P2W7")
)

my_expt <- tibble(
  value = runif(6),
  sample = c("control1", "Control2", "cONtrol3", "treated1", "Treated2", "TREATed3")
)

# Use `protein_small` df for the following examples -
# Only keep proteins whose name begins with an "A"


str_view_all(protein_small$protein_names, "A")
str_view_all(protein_small$protein_names, "^A")

# Remove "_HUMAN" from protein names from the `protein_small` data frame
protein_small <- protein_small %>% 
  mutate(gene_names = str_replace(protein_names, "_HUMAN", ""))

# Use `my_expt` df for the following examples -
# Convert all sample entries to lower case
my_expt <- my_expt %>% 
  mutate(sample = str_to_lower(sample))

# Create a new column called "condition" that contains the sample entries with the numbers removed

```

# Advanced Regex

```{r}
# REMOVE EXTRA INFO FROM GENES (several ways to do it)
gene_names <- c("A1BG (1)", "A1CF (29974)", "A2M (2)", "A2ML1 (144568)", "A3GALT2 (127550)", "A4GALT (53947)", "A4GNT (51146)", "AAAS (8086)", "AACS (65985)")

# Remove white spaces, parenthesis, and everything in the parenthesis
gene_names %>%
  stringr::str_replace("\\s*\\([^\\)]+\\)", "")
# Use the Stringr cheatsheet (page 2) to work through this complex example!

# NOTE: All special characters must be preceded by a \
# Example -
# How would you detect only the first element of this vector?
a <- c("ab.", "abc")
str_detect(a, ".") # we detect both!

# We must use \. to tell R we literally want a .
# To do that we use the special character \\.
str_detect(a, "\\.") # this works!

# Remove everything after a white space, including the white space
gene_names %>% 
  stringr::str_remove("\\s.*")

# Compute the START position and the END position of the FIRST "word" of the string using a white space as a separator between words by default. Then, subset and return everything in this position range, that is, the first "word"
stringr::word(gene_names)

# Demonstration that this function is returning the first "word"
stringr::word("A1BG A1CF")

# Demonstration that this function is using a white space as a separator between words by default
# This is returning everything before "1" as a "word"
stringr::word(gene_names, sep = "1") # default is 'sep = fixed(" ")'
```

# Cleaning up column names

```{r}
# Example data frame with messy column names
typical_excel <- tibble(
`Gene Names` = c("A1BG_HUMAN", "A1CF_HUMAN", "B3GAT1_HUMAN"),
`uniprot id` = c("P04217", "Q9NQ94", "Q9P2W7")
)
# Why are the column names of this data frame problematic?
typical_excel %>% 
  filter(Gene Names == "A1BG_HUMAN") # expect error

# Make the column names more "programming friendly"
typical_r <- janitor::clean_names(typical_excel)

# Replace underscores with periods in column names (this is common in a lot of bioinformatics scripts)
colnames(typical_r) <- str_replace(colnames(typical_r), "_", ".")
```

